### PR TITLE
Cleanup util/chplenv/third-party-pkgs

### PR DESCRIPTION
**Pull request by ronawho**
Make it more pythonic and cleanup implementation:
 - don't hardcode '/', use os.path.join
 - don't use file open/close, use with stmt
 - don't manually look at sys.argv, use optparse
 - use list comprehension to cleanup compiler/runtime filtering
 - `list.count(val)` => `val in list`
 - don't recompute chpl_home/third-party, use chpl_home_utils
 - split code into functions
 - use run_command instead of subprocess+decode
 - better var names, and avoid using builtin names like "dir"

### CODE REVIEW

**Comment by ronawho** [comment]
Probably easiest to review side-by-side: https://github.com/chapel-lang/chapel/pull/11709/files?utf8=%E2%9C%93&diff=split&w=1

Or by looking at the before and after side-by-side: 
 - https://github.com/chapel-lang/chapel/blob/0a5ac4ee36/util/chplenv/third-party-pkgs
 - https://github.com/chapel-lang/chapel/blob/22e4c8587c/util/chplenv/third-party-pkgs

File: util/chplenv/third-party-pkgs [hunk 1]
```diff
@@ -1,92 +1,107 @@
...
+   - Any Chapel environment variable ending in '_PLATFORM' describes
+     the hardware/OS combo and will not considered in this script
+   - "Top-level" Chapel environment variables are printed starting
+     in the first column.  Any Chapel environment variables that 
+     indented are assumed to be modifiers of the corresponding top-level
+     variable and will not be considered in this script (*** and also
+     do not necessarily have to be prefixed with 'CHPL_')
+ - $CHPL_HOME/third-party/<pkg>/Makefile.include defines any aliases
+   for the package name in a variable called XXX_ALIASES where XXX
+   is the directory name in all caps.
```
Comment (mppf) [inline util/chplenv/third-party-pkgs]: Let's also point out here that this must use = and consider having an example.
```
This alias must be defined with = assignment. For example:
    FOOLIB_ALIASES = libfoo
```

File: util/chplenv/third-party-pkgs [hunk 2]
```diff
@@ -1,92 +1,107 @@
...
-
-pkgs_text = ' '.join(pkgs)
-
-print(pkgs_text)
 
+"""
+ Print the names of any third-party packages needed for the current
+ Chapel environment settings
+
+ ASSUMPTIONS:
```
Comment (ben-albrecht) [inline util/chplenv/third-party-pkgs]: Aren't a lot of these assumptions made in all `chplenv` scripts? I find it odd that only this one documents them. Perhaps the common assumptions should be put into a local `README.md` or the `__init__.py`?

File: util/chplenv/third-party-pkgs [hunk 3]
```diff
@@ -1,92 +1,107 @@
...
+def gather_chplenv():
+    """ Run printchplenv and return the output """
+    printchplenv = os.path.join(get_chpl_home(), 'util', 'printchplenv')
+    args = ['--simple', '--internal', '--runtime', '--anonymize', '--no-tidy']
+    return utils.run_command([printchplenv] + args)
+
+def compute_required_pkgs():
+    """ Determine required third-party packages based on the chplenv """
+    chplenv = gather_chplenv()
+    pkg_aliases = find_pkg_alises()
```
Comment (ben-albrecht) [inline util/chplenv/third-party-pkgs]: aliases

File: util/chplenv/third-party-pkgs [hunk 4]
```diff
@@ -1,92 +1,107 @@
...
+import os
+import sys
+
+chplenv_dir = os.path.dirname(__file__)
+sys.path.insert(0, os.path.abspath(chplenv_dir))
+
+import utils
+from chpl_home_utils import get_chpl_third_party, get_chpl_home
+
+def get_all_pkgs():
```
Comment (ben-albrecht) [inline util/chplenv/third-party-pkgs]: micro-optimizations (therefore, feel free to ignore):

- Making this a set would give O(1) lookup times (assuming no duplicates and order is not important).
- Memoizing this or storing it into a global variable would also reduce unnecessary `os.listdir()` calls.

Comment (ronawho) [reply util/chplenv/third-party-pkgs]: This is only called a single time from the runtime Makefile (i.e only called during `make`) not during regular chplenv/printchplenv invocations so I'm not worried about micro-optimizations here.

**Comment by ronawho** [comment]
> How's our pylint score?

8.5, but remaining complaints are not very helpful. I'm not going to rename this script even though it's not good convention, we're stuck with optparse, f is a fine variable name for files, and utils most certainly has a run_command.

```
************* Module chplenv.third-party-pkgs
util/chplenv/third-party-pkgs:1:0: C0103: Module name "third-party-pkgs" doesn't conform to snake_case naming style (invalid-name)
util/chplenv/third-party-pkgs:28:0: W0402: Uses of a deprecated module 'optparse' (deprecated-module)
util/chplenv/third-party-pkgs:47:44: C0103: Variable name "f" doesn't conform to snake_case naming style (invalid-name)
util/chplenv/third-party-pkgs:61:11: E1101: Module 'utils' has no 'run_command' member (no-member)

------------------------------------------------------------------
Your code has been rated at 8.55/10 (previous run: 8.55/10, +0.00)
```

File: util/chplenv/third-party-pkgs (hunk 4 above)
Comment (ben-albrecht) [reply util/chplenv/third-party-pkgs]: I did not realize that. Sounds good.

**Comment by ronawho** [comment]
Just fyi @mppf CHPL_LLVM is not included in the printchplenv command being used, so `$CHPL_HOME/util/chplenv/third-party-pkgs --compiler` will never print anything now. That's true on master too, not a change in this PR. Presumably that's not an issue and we trigger off of something else for the llvm build now?

**Comment by mppf** [comment]
@ronawho - I wouldn't bother with getting LLVM to work here but I don't object to whatever you do. The LLVM build/Makefile is directly depended upon in the compiler/Makefile. We only run third-party-pkgs for the runtime make, so isn't it moot unless you also to rearrange how compiler/Makefile depends on LLVM?

I'd be happy to have this differentiate with `printchplenv --runtime` vs `printchplenv --compiler` but the main concern in third-party-pkgs with LLVM is that it shouldn't be a dependency for the *runtime*.

**Comment by ronawho** [comment]
Sounds good to me. For now I just added `--compiler` to the printchplenv command for now and verified that we don't print out `llvm` for non-compiler runs of third-party-pkgs. I thought about just removing `third-party-pkgs --compiler` support, but I'll save that or any other refactoring for a later day.

**Pull request merged**